src/main/java/org/broadinstitute/hellbender/utils/hmm/segmentation/HiddenMarkovModelSegmentProcessor.java

```diff
@@ -0,0 +1,652 @@
+import org.broadinstitute.hellbender.utils.QualityUtils;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.hmm.ForwardBackwardAlgorithm;
+import org.broadinstitute.hellbender.utils.hmm.interfaces.CallStringProducer;
+import org.broadinstitute.hellbender.utils.hmm.interfaces.ScalarProducer;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.*;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * TODO github/gatk-protected issue #855 -- this is a last-minute fork of {@link HiddenMarkovModelPostProcessor}.
+ * It just does segmentation (no VCF creation)
+ *
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+public class HiddenMarkovModelSegmentProcessor<D, S extends CallStringProducer & ScalarProducer, T extends Target> {
+
+    private final Logger logger = LogManager.getLogger(HiddenMarkovModelSegmentProcessor.class);
+
+    /**
+     * Maximum tolerated log probability value. Values between this constant and 0.0 as considered as a 0.0.
+     * Values above this threshold are considered to indicate a log probability calculation problem that is
+     * worth to be reported as an error or warning.
+     */
+    public static final double MAX_LOG_PROB = 1e-3;
+
+    /**
+     * Maximum reportable output quality score; higher quality scores will
+     * capped to this value.
+     */
+    public static final double MAX_QUAL_SCORE = 199.99999999999;
+
+    /**
+     * Precision of Phred-scale scores
+     */
+    public static final double PHRED_SCORE_PRECISION = 0.1;
+
+    /**
+     * Calculating variance posterior on long segments is time consuming; we assume hidden states become
+     * independent if their separation is larger than the following value
+     */
+    public static final double INDEPENDENT_TARGETS_SEPARATION_THRESHOLD = 0;
+
+    /**
+     * Threshold used to determine best way to calculate log(1- exp(a))
+     * based on https://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf
+     */
+    private static final double LN_1_M_EXP_THRESHOLD = - Math.log(2);
+
+    /**
+     * Cached value of ln(10)^-1
+     */
+    private static final double INV_LN_10 = 1.0 / Math.log(10);
+
+    private final List<String> sampleNames;
+    private final List<SexGenotypeData> sampleSexGenotypes;
+    private final BiFunction<SexGenotypeData, Target, S> referenceStateFactory;
+    private final int numSamples;
+    private final List<TargetCollection<T>> sampleTargets;
+    private final List<ForwardBackwardAlgorithm.Result<D, T, S>> sampleForwardBackwardResults;
+    private final List<List<S>> sampleBestPaths;
+
+    /**
+     * An intermediate result of segment calling -- will be used for VCF creation if an external
+     * segments file is not given
+     */
+    private Map<String, List<HiddenStateSegment<S, T>>> allSegmentsBySampleName = null;
```

Comment:
I think this can be `final` or made a local variable in the appropriate method.

Is there a good reason it was allowed to be `null` in `HiddenMarkovModelPostProcessor`?  I guess you might want to output the segmentation multiple times and avoid recomputing it, as well as create a `HiddenMarkovModelPostProcessor` without necessarily calculating the segmentation, but I don't know if this happens in practice and it all feels quite awkward.